### PR TITLE
add flake attribute apps to make it easier to run

### PR DESCRIPTION
**Pull request by contrun**
### Checklist

- [ ] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [ ] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -30,6 +30,15 @@
...
         forAllSystems (system: self.packages.${system}.home-manager);
 
+      apps = forAllSystems (system: {
+        home-manager = {
+          type = "app";
+          program = "${defaultPackage."${system}"}/bin/home-manager";
+        };
+      });
+
+      defaultApp = forAllSystems (system: apps."${system}".home-manager);
```
Comment (teto) [inline flake.nix:40]: you dont need the quotes.

Comment (contrun) [reply flake.nix]: Removed

**Review by berbiche — CHANGES_REQUESTED** [review]
Thanks for the PR!

Can you remove the committed `flake.lock` file?

**Comment by contrun** [comment]
@berbiche It seems `nix run` requires a `flake.lock`. What's the downside of including a `flake.lock`?

**Comment by berbiche** [comment]
> @berbiche It seems `nix run` requires a `flake.lock`.

`--no-write-lock-file`

> What's the downside of including a `flake.lock`?

The downside is that we would need to keep HM's flake inputs updated and we currently do not have an automated way of doing it.
There's a PR to add a workflow to update HM's flake.lock that is still waiting for something (I lost track).

**Comment by contrun** [comment]
@berbiche Maybe you are referring to https://github.com/nix-community/home-manager/pull/1913 ?

**Comment by berbiche** [comment]
> @berbiche Maybe you are referring to #1913 ? I removed flake.lock for now.

https://github.com/nix-community/home-manager/pull/1934

**Review by berbiche — APPROVED** [review]

**Comment by terlar** [comment]
@contrun thanks for this, what do you think about having the defaultApp be flake native, e.g. add `--flake` to the command first. We could provide another app that returns home-manager without the flake flag.

I mentioned this earlier here:
https://github.com/nix-community/home-manager/issues/2523#issuecomment-983964840

With such a wrapper we could have:
```
nix run github:nix-community/home-manager switch .
```

And for the non-flake one it could be something like:
```
nix run github:nix-community/home-manager#legacy switch
```

**Comment by contrun** [comment]
@terlar I agree. Currently we can use flake for home-manager by passing `--flake flake-uri` to home-manager.  Users may expect to run home-manager like `home-manager switch [flake-uri]`, where `flake-uri` is optional. I don't know if wrapping `home-manager` with `home-manager --flake` is acceptable for now, or we need to change the way home-manager taking arguments. Maybe home-manager can have a feature gate, i.e. if environment variable `USE_FLAKE` is present, then home-manager should be flake-native.

**Comment by akavel** [comment]
> (...) Maybe home-manager can have a feature gate (...)

Noob question (I reported the #2523, so I'm an interested party noob): could it make sense to do the detection based on whether a `flake.nix` file exists at the target location (i.e. `~/.config/nixpkgs/flake.nix` by default, or `[flake-uri]/flake.nix`, or whatever)?

To my noobish eyes this sounds reasonable: if someone has a `flake.nix` file, the default assumption that they want to opt-in to flakes seems to make sense (and could still be e.g. opted out with some `--no-flake` flag possibly); OTOH, if someone doesn't have `flake.nix`, they probably want the old behavior by default.

**Comment by contrun** [comment]
@akavel All the configuration should be contained in the flake itself. That's the whole point of nix flakes, to make the building hermetic. Flake version of home-manager shouldn't depend on host-specific behavior.

**Comment by ncfavier** [comment]
Why was this added? The default behaviour of `nix run` (since at least 2.4) if there are no apps is to run `/bin/${pname}` under the default package, which does what we want here.

Adding an app would only be useful if we wanted to add a default `--flake` flag as suggested, but I don't think it's a good idea to artificially introduce CLI differences based on how the command is invoked. Instead, we could make `home-manager` take an optional flake as a positional argument: `home-manager [OPTIONS...] COMMAND [FLAKE]`.

Also note that this stopped working in Nix 2.8: https://github.com/NixOS/nix/issues/6448#issuecomment-1132855605

**Pull request merged**